pkg/operator/controller/garden/access/reconciler.go

```diff
@@ -0,0 +1,132 @@
+	log := logf.FromContext(ctx)
+
+	secret := &corev1.Secret{}
+	if err := r.Client.Get(ctx, request.NamespacedName, secret); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.V(1).Info("Object is gone, stop reconciling")
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("error retrieving object from store: %w", err)
+	}
+
+	kubeConfigRaw, ok := secret.Data[kubernetesclient.KubeConfig]
+	if !ok {
+		log.Info("Secret does not contain kubeconfig")
+		return reconcile.Result{}, nil
+	}
+
+	restConfig, err := clientcmd.RESTConfigFromKubeConfig(kubeConfigRaw)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("error creating REST config: %w", err)
+	}
+
+	if restConfig.BearerToken == "" {
+		log.Info("BearerToken is not yet populated in kubeconfig")
+		return reconcile.Result{}, nil
+	}
+
+	if err := r.writeTokenToFile(restConfig.BearerToken); err != nil {
+		return reconcile.Result{}, fmt.Errorf("error writing bearer token to file: %w", err)
+	}
+
+	return reconcile.Result{}, r.createVirtualCluster(log, restConfig)
+}
+
+// GetVirtualCluster returns the virtual cluster object.
+func (r *Reconciler) GetVirtualCluster() cluster.Cluster {
+	return r.virtualCluster
+}
+
+// CreateTemporaryFile creates a temporary file. Exposed for testing.
+var CreateTemporaryFile = afero.TempFile
+
+func (r *Reconciler) writeTokenToFile(token string) error {
+	if r.tokenFilePath == "" {
+		tokenFile, err := CreateTemporaryFile(r.FS, "", "garden-access")
+		if err != nil {
+			return fmt.Errorf("error creating gardener-access-kubeconfig file: %w", err)
+		}
+		r.tokenFilePath = tokenFile.Name()
+		if err := tokenFile.Close(); err != nil {
+			return fmt.Errorf("error closing gardener-access-kubeconfig file: %w", err)
+		}
+	}
+
+	return afero.WriteFile(r.FS, r.tokenFilePath, []byte(token), 0o600)
+}
+
+func (r *Reconciler) createVirtualCluster(log logr.Logger, restConfig *rest.Config) error {
+	if r.virtualCluster != nil {
+		return nil
+	}
+
+	// prepare REST config
+	conf := &kubernetesclient.Config{}
+	if err := kubernetesclient.WithRESTConfig(restConfig)(conf); err != nil {
+		return fmt.Errorf("error setting rest config: %w", err)
+	}
+	if err := kubernetesclient.WithClientConnectionOptions(r.Config.VirtualClientConnection)(conf); err != nil {
+		return fmt.Errorf("error setting client connection configuration: %w", err)
+	}
+	restConfig.BearerToken = ""
+	restConfig.BearerTokenFile = r.tokenFilePath
+
+	virtualCluster, err := cluster.New(restConfig, func(opts *cluster.Options) {
+		opts.Scheme = operatorclient.VirtualScheme
+		opts.Logger = log
+	})
+	if err != nil {
+		return fmt.Errorf("could not instantiate virtual cluster: %w", err)
+	}
+	if err := r.Manager.Add(virtualCluster); err != nil {
+		return fmt.Errorf("failed adding virtual cluster to manager: %w", err)
+	}
+
+	r.virtualCluster = virtualCluster
+
+	return nil
+}
```

Comment:
I really don't want to sabotage you, but I still find it a but unintuitive that this controller creates the `cluster.Cluster` object. I see and agree that it is simple and straight-forward, but as reader of the code, I wouldn't expect it here.
What I have in mind is that we decide on a temporary file path before instantiating the `garden-access` controller, and pass this path to it. This way, another controller could just wait for the token to be written to the path, and then create and add the `cluster.Cluster`.

Do you agree that this makes sense conceptually or do you see any difficulties/downsides with it? I leave it up to you if you want to follow-up with it. We could also consider it for a subsequent improvement/PR after this one got merged.